apt-hook/json-hook.go

```diff
@@ -0,0 +1,628 @@
+package main
+
+import (
+	"bufio"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net"
+	"os"
+	"strconv"
+)
+
+type jsonRPCPackageVersion struct {
+	Id           int    `json:"id"`
+	Version      string `json:"version"`
+	Architecture string `json:"architecture"`
+	Pin          int    `json:"pin"`
+	Origins      []struct {
+		Archive  string `json:"archive"`
+		Codename string `json:"codename"`
+		Version  string `json:"version"`
+		Origin   string `json:"origin"`
+		Label    string `json:"label"`
+		Site     string `json:"site"`
+	} `json:"origins"`
+}
+type jsonRPC struct {
+	JsonRPC string `json:"jsonrpc"`
+	Method  string `json:"method"`
+	Params  struct {
+		Command         string   `json:"command"`
+		UnknownPackages []string `json:"unknown-packages"`
+		Packages        []struct {
+			Id           int    `json:"id"`
+			Name         string `json:"name"`
+			Architecture string `json:"architecture"`
+			Mode         string `json:"mode"`
+			Automatic    bool   `json:"automatic"`
+			Versions     struct {
+				Candidate jsonRPCPackageVersion `json:"candidate"`
+				Install   jsonRPCPackageVersion `json:"install"`
+				Current   jsonRPCPackageVersion `json:"current"`
+			} `json:"versions"`
+		} `json:"packages"`
+	} `json:"params"`
+}
+
+func createUpdateMessage(standardSecurityCount int, esmAppsCount int, esmInfraCount int) string {
+	displayStandard := true
+	// TODO other lts/beta criteria for displaying this?
+	displayEsmApps := true
+	displayEsmInfra := true
+
+	if standardSecurityCount == 0 {
+		displayStandard = false
+	}
+	if esmAppsCount == 0 {
+		displayEsmApps = false
+	}
+	if esmInfraCount == 0 {
+		displayEsmInfra = false
+	}
+
+	// There are probably more clever ways of breaking down these messages, but this makes
+	// it easy to follow the minor variations
+	if displayStandard && displayEsmApps && displayEsmInfra {
```

Comment:
I think I have an idea to make the code a little simpler. We could built each service message separately and then just figure out how to connect them. My idea would be like this (Sorry, will use python here because I am go noob)

```python
standard_msg = build_standard_message(standardSecurityCount) # x standard security update
esm_infra_msg = build_esm_infra_message(esmInfraCount)  # x esm-infra update
esm_apps_msg = build_esm_apps_message(esmAppsCount)  # x esm-apps update

if not standard_message_msg
    standard_msg_unifier = ""
elif esm_infra_msg and esm_apps_message:
    standard_msg_unifier = ", "
elif esm_infra_msg or esm_apps_message:
    standard_msg_unifier = " and "
else:
    standard_msg_unifier = ""

if standard_msg_unifier == ", ":
    esm_unifier = " and "
elif standard_msg_unifier == " and ":
    esm_unifier = ""
elif esm_infra_msg and esm_apps_msg:
   esm_unifier = "and"

return standard_security_msg + standard_msg_unifier + esm_infra_msg + esm_msg_unifier + esm_apps_msg
```

Reply:
I think the logic will hold, but maybe there are gaps on it that I could not see

Reply:
Thanks for this! I updated the logic to construct the message in a similar manner to what you suggest here.